src/picongpu/include/particles/ionization/byField/BSI/BSI_Impl.hpp

```diff
@@ -0,0 +1,195 @@
+/**
+ * Copyright 2015 Marco Garten
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "types.h"
+#include "math/vector/Size_t.hpp"
+#include "simulation_defines.hpp"
+#include "traits/Resolve.hpp"
+#include "mappings/kernel/AreaMapping.hpp"
+
+#include "fields/FieldB.hpp"
+#include "fields/FieldE.hpp"
+
+#include "particles/ionization/byField/BSI/BSI.def"
+#include "particles/ionization/byField/BSI/AlgorithmBSI.hpp"
+#include "particles/ionization/ionization.hpp"
+
+#include "compileTime/conversion/TypeToPointerPair.hpp"
+#include "memory/boxes/DataBox.hpp"
+
+#include "particles/ParticlesFunctors.hpp"
+
+namespace picongpu
+{
+namespace particles
+{
+namespace ionization
+{
+
+    /** \struct BSI_Impl
+     * \brief Barrier Suppression Ionization - Implementation
+     *
+     * \tparam T_DestSpecies electron species to be created
+     * \tparam T_SrcSpecies particle species that is ionized
+     */
+    template<typename T_DestSpecies, typename T_SrcSpecies>
+    struct BSI_Impl
+    {
+
+        typedef T_DestSpecies DestSpecies;
+        typedef T_SrcSpecies  SrcSpecies;
+
+        typedef typename SrcSpecies::FrameType FrameType;
+
+        /* specify field to particle interpolation scheme */
+        typedef typename PMacc::traits::Resolve<
+            typename GetFlagType<FrameType,interpolation<> >::type
+        >::type InterpolationScheme;
+
+        /* margins around the supercell for the interpolation of the field on the cells */
+        typedef typename GetMargin<InterpolationScheme>::LowerMargin LowerMargin;
+        typedef typename GetMargin<InterpolationScheme>::UpperMargin UpperMargin;
+
+        /* relevant area of a block */
+        typedef SuperCellDescription<
+            typename MappingDesc::SuperCellSize,
+            LowerMargin,
+            UpperMargin
+            > BlockArea;
```

Comment:
Some algorithms needs more cells in one direction than in the other. e.g. interpolation needs `shape/2` LowerMargin and `shape/2+1` UpperMargin. The difference come from the fact that our in cell origin is in 2D on the left top corner, this means that our coordinate system needs mostly one cell more in positive (upper) than in negative (lower) direction.